Check `GH_REPO` too in addition to `--repo` for disambiguation

williammartin opened the pull request:
## Description

Fixes #10525

Commit pulled from @iamazeem's fork, with attribution maintained.

This is an alternative to https://github.com/cli/cli/pull/10535 because that one requires a bit more thought and I'd like this to go out in a release.

## Acceptance

**Given** my cwd is not a git repo
**And** I have `GH_REPO` set to a repo I have permissions on
**When** I run any of the `gh secret` commands
**Then** I am able to interact with secrets on that repo

Before:

```
➜  workspace GH_REPO=williammartin-test-org/test-repo gh secret list
failed to run git: fatal: not a git repository (or any of the parent directories): .git
```

After:

```
➜  workspace GH_REPO=williammartin-test-org/test-repo ~/workspace/cli/bin/gh secret list | cat
FOO     2025-01-09T16:39:07Z
```

iamazeem committed:
[gh secret] Check `GH_REPO` too in addition to `--repo` for disambiguation

andyfeller commented:
## Confirming changes with updated tests

Following up on this PR, I'm on the verge of pushing up changes to tests to cover this and want to demonstrate that these tests fail on `trunk` before demonstrating they pass on the branch:

```shell
Running tool: /opt/homebrew/bin/go test -timeout 30s -run ^TestNewCmdListBaseRepoFuncs$ github.com/cli/cli/v2/pkg/cmd/secret/list

--- FAIL: TestNewCmdListBaseRepoFuncs (0.00s)
    --- FAIL: TestNewCmdListBaseRepoFuncs/when_GH_REPO_env_var_is_provided,_the_factory_base_repo_func_is_used (0.00s)
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x2 addr=0x28 pc=0x1032fdec0]

goroutine 6 [running]:
testing.tRunner.func1.2({0x10358c3a0, 0x1039e9a50})
	/opt/homebrew/Cellar/go/1.24.0/libexec/src/testing/testing.go:1734 +0x1ac
testing.tRunner.func1()
	/opt/homebrew/Cellar/go/1.24.0/libexec/src/testing/testing.go:1737 +0x334
panic({0x10358c3a0?, 0x1039e9a50?})
	/opt/homebrew/Cellar/go/1.24.0/libexec/src/runtime/panic.go:787 +0x124
github.com/cli/cli/v2/internal/ghrepo.IsSame({0x10362d828, 0x140001da7b0}, {0x0, 0x0})
	/Users/andyfeller/Documents/workspace/cli/cli-10539/internal/ghrepo/repo.go:80 +0x40
github.com/cli/cli/v2/pkg/cmd/secret/list.TestNewCmdListBaseRepoFuncs.func2(0x140001ea380)
	/Users/andyfeller/Documents/workspace/cli/cli-10539/pkg/cmd/secret/list/list_test.go:217 +0x3f8
testing.tRunner(0x140001ea380, 0x140001d37a0)
	/opt/homebrew/Cellar/go/1.24.0/libexec/src/testing/testing.go:1792 +0xe4
created by testing.(*T).Run in goroutine 4
	/opt/homebrew/Cellar/go/1.24.0/libexec/src/testing/testing.go:1851 +0x374
FAIL	github.com/cli/cli/v2/pkg/cmd/secret/list	0.208s
FAIL


Running tool: /opt/homebrew/bin/go test -timeout 30s -run ^TestNewCmdDeleteBaseRepoFuncs$ github.com/cli/cli/v2/pkg/cmd/secret/delete

--- FAIL: TestNewCmdDeleteBaseRepoFuncs (0.00s)
    --- FAIL: TestNewCmdDeleteBaseRepoFuncs/when_GH_REPO_env_var_is_provided,_the_factory_base_repo_func_is_used (0.00s)
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x2 addr=0x28 pc=0x104758000]

goroutine 37 [running]:
testing.tRunner.func1.2({0x104993fa0, 0x104dac9d0})
	/opt/homebrew/Cellar/go/1.24.0/libexec/src/testing/testing.go:1734 +0x1ac
testing.tRunner.func1()
	/opt/homebrew/Cellar/go/1.24.0/libexec/src/testing/testing.go:1737 +0x334
panic({0x104993fa0?, 0x104dac9d0?})
	/opt/homebrew/Cellar/go/1.24.0/libexec/src/runtime/panic.go:787 +0x124
github.com/cli/cli/v2/internal/ghrepo.IsSame({0x104a30788, 0x140001fe6f0}, {0x0, 0x0})
	/Users/andyfeller/Documents/workspace/cli/cli-10539/internal/ghrepo/repo.go:80 +0x40
github.com/cli/cli/v2/pkg/cmd/secret/delete.TestNewCmdDeleteBaseRepoFuncs.func2(0x14000129180)
	/Users/andyfeller/Documents/workspace/cli/cli-10539/pkg/cmd/secret/delete/delete_test.go:238 +0x3f8
testing.tRunner(0x14000129180, 0x140001f3810)
	/opt/homebrew/Cellar/go/1.24.0/libexec/src/testing/testing.go:1792 +0xe4
created by testing.(*T).Run in goroutine 35
	/opt/homebrew/Cellar/go/1.24.0/libexec/src/testing/testing.go:1851 +0x374
FAIL	github.com/cli/cli/v2/pkg/cmd/secret/delete	0.253s
FAIL


Running tool: /opt/homebrew/bin/go test -timeout 30s -run ^TestNewCmdSetBaseRepoFuncs$ github.com/cli/cli/v2/pkg/cmd/secret/set

--- FAIL: TestNewCmdSetBaseRepoFuncs (0.00s)
    --- FAIL: TestNewCmdSetBaseRepoFuncs/when_GH_REPO_env_var_is_provided,_the_factory_base_repo_func_is_used (0.00s)
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x2 addr=0x28 pc=0x100ac8a10]

goroutine 6 [running]:
testing.tRunner.func1.2({0x100d14aa0, 0x101138b40})
	/opt/homebrew/Cellar/go/1.24.0/libexec/src/testing/testing.go:1734 +0x1ac
testing.tRunner.func1()
	/opt/homebrew/Cellar/go/1.24.0/libexec/src/testing/testing.go:1737 +0x334
panic({0x100d14aa0?, 0x101138b40?})
	/opt/homebrew/Cellar/go/1.24.0/libexec/src/runtime/panic.go:787 +0x124
github.com/cli/cli/v2/internal/ghrepo.IsSame({0x100db2cb8, 0x140001d8810}, {0x0, 0x0})
	/Users/andyfeller/Documents/workspace/cli/cli-10539/internal/ghrepo/repo.go:80 +0x40
github.com/cli/cli/v2/pkg/cmd/secret/set.TestNewCmdSetBaseRepoFuncs.func2(0x140001ea380)
	/Users/andyfeller/Documents/workspace/cli/cli-10539/pkg/cmd/secret/set/set_test.go:337 +0x3f8
testing.tRunner(0x140001ea380, 0x140001e22a0)
	/opt/homebrew/Cellar/go/1.24.0/libexec/src/testing/testing.go:1792 +0xe4
created by testing.(*T).Run in goroutine 4
	/opt/homebrew/Cellar/go/1.24.0/libexec/src/testing/testing.go:1851 +0x374
FAIL	github.com/cli/cli/v2/pkg/cmd/secret/set	0.248s
FAIL
```

andyfeller committed:
Expand gh secret base repo tests

Building on top of the work done in commands to account for GH_REPO environment variable, this commit expands existing tests around handling base repo functions to include new test scenarios.

These tests fail in the same way as reported on the issue when run against `trunk` without the other branch's changes, demonstrating they will help avoid regression.

andyfeller committed:
Expand gh secret acceptance tests

This commit builds on top of work previously done within acceptance tests around gh secret and remote disambiguation.

The choice of expanding this existing test rather than creating a new test was in order to keep context together within a single acceptance test rather than splitting it over multiple.

andyfeller commented:
## Updated acceptance test results

Along with the updated tests, I've gone ahead and enhanced our acceptance tests to also cover this.

Here is the result from running the updated acceptance tests against `trunk` failing to handle `GH_REPO`:

```shell
‹trunk*›$ set -o pipefail && GH_ACCEPTANCE_SCRIPT=secret-require-remote-disambiguation.txtar GH_ACCEPTANCE_TOKEN=$(gh auth token --hostname github.com) GH_ACCEPTANCE_HOST=github.com GH_ACCEPTANCE_ORG=gh-acceptance-testing go test -tags acceptance -json -run ^TestSecrets$ github.com/cli/cli/v2/acceptance  | tparse --all go test
┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┓
┃   FAIL  package: github.com/cli/cli/v2/acceptance   ┃
┗━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┛

--- FAIL: TestSecrets (0.00s)
--- FAIL: TestSecrets/secret-require-remote-disambiguation (7.84s)

    testscript.go:584: WORK=$WORK
        PATH=/var/folders/y5/q89b76cj7ws55x3trcyllf_80000gn/T/testscript-main3772044479/bin:/opt/homebrew/Cellar/go/1.24.0/libexec/bin:/Users/andyfeller/.nvm/versions/node/v18.17.1/bin:/usr/local/opt/go/libexec/bin:/opt/homebrew/bin:/Users/andyfeller/go/bin:/Users/andyfeller/.dotnet:/Users/andyfeller/bin:/Users/andyfeller/.rbenv/shims:/opt/homebrew/bin:/opt/homebrew/sbin:/usr/local/bin:/System/Cryptexes/App/usr/bin:/usr/bin:/bin:/usr/sbin:/sbin:/var/run/com.apple.security.cryptexd/codex.system/bootstrap/usr/local/bin:/var/run/com.apple.security.cryptexd/codex.system/bootstrap/usr/bin:/var/run/com.apple.security.cryptexd/codex.system/bootstrap/usr/appleinternal/bin:/usr/local/share/dotnet:~/.dotnet/tools
        GOTRACEBACK=system
        HOME=$WORK
        TMPDIR=$WORK/.tmp
        devnull=/dev/null
        /=/
        :=:
        $=$
        exe=
        SCRIPT_NAME=secret_require_remote_disambiguation
        GH_CONFIG_DIR=$WORK
        GH_HOST=github.com
        ORG=gh-acceptance-testing
        GH_TOKEN=gho_************************************
        RANDOM_STRING=pTzaHEGHvk
        
        # Set up env vars (0.000s)
        > env REPO=${SCRIPT_NAME}-${RANDOM_STRING}
        # Use gh as a credential helper (0.560s)
        > exec gh auth setup-git
        # Create a repository with a file so it has a default branch (0.928s)
        > exec gh repo create ${ORG}/${REPO} --add-readme --private
        [stdout]
        https://github.com/gh-acceptance-testing/secret_require_remote_disambiguation-pTzaHEGHvk
        # Defer repo cleanup (0.000s)
        > defer gh repo delete --yes ${ORG}/${REPO}
        # Create a fork (0.849s)
        > exec gh repo fork ${ORG}/${REPO} --org ${ORG} --fork-name ${REPO}-fork
        [stdout]
        https://github.com/gh-acceptance-testing/secret_require_remote_disambiguation-pTzaHEGHvk-fork
        # Defer fork cleanup (0.000s)
        > defer gh repo delete --yes ${ORG}/${REPO}-fork
        # Sleep to allow the fork to be created before cloning (2.001s)
        > sleep 2
        # Clone and move into the fork repo (1.514s)
        > exec gh repo clone ${ORG}/${REPO}-fork
        [stderr]
        Cloning into 'secret_require_remote_disambiguation-pTzaHEGHvk-fork'...
        From https://github.com/gh-acceptance-testing/secret_require_remote_disambiguation-pTzaHEGHvk
         * [new branch]      main       -> upstream/main
        > cd ${REPO}-fork
        $WORK/secret_require_remote_disambiguation-pTzaHEGHvk-fork
        # Secret list requires disambiguation (0.071s)
        > ! exec gh secret list
        [stderr]
        multiple remotes detected. please specify which repo to use by providing the -R, --repo argument
        [exit status 1]
        > stderr 'multiple remotes detected. please specify which repo to use by providing the -R, --repo argument'
        # Secret set requires disambiguation (0.051s)
        > ! exec gh secret set 'TEST_SECRET_NAME' --body 'TEST_SECRET_VALUE'
        [stderr]
        multiple remotes detected. please specify which repo to use by providing the -R, --repo argument
        [exit status 1]
        > stderr 'multiple remotes detected. please specify which repo to use by providing the -R, --repo argument'
        # Secret delete requires disambiguation (0.049s)
        > ! exec gh secret delete 'TEST_SECRET_NAME'
        [stderr]
        multiple remotes detected. please specify which repo to use by providing the -R, --repo argument
        [exit status 1]
        > stderr 'multiple remotes detected. please specify which repo to use by providing the -R, --repo argument'
        # Move out of the fork repo to test whether secret commands work without local repository context (0.000s)
        > cd ..
        $WORK
        # Secret set using --repo flag does not require disambiguation (0.370s)
        > exec gh secret set 'TEST_SECRET_NAME' --body 'TEST_SECRET_VALUE' --repo ${ORG}/${REPO}-fork
        # Secret list using --repo flag does not require disambiguation (0.264s)
        > exec gh secret list --repo ${ORG}/${REPO}-fork
        [stdout]
        TEST_SECRET_NAME	2025-03-04T19:27:33Z
        > stdout 'TEST_SECRET_NAME'
        # Secret delete using --repo flag does not require disambiguation (0.274s)
        > exec gh secret delete 'TEST_SECRET_NAME' --repo ${ORG}/${REPO}-fork
        # Setup GH_REPO for testing environment variable behavior (0.000s)
        > env GH_REPO=${ORG}/${REPO}-fork
        # Secret set using GH_REPO env var does not require disambiguation (0.866s)
        > exec gh secret set 'TEST_SECRET_NAME2' --body 'TEST_SECRET_VALUE2'
        [stderr]
        failed to run git: fatal: not a git repository (or any of the parent directories): .git
        
        [exit status 1]
        FAIL: testdata/secret/secret-require-remote-disambiguation.txtar:55: unexpected command failure
        

┌────────────────────────────────────────────────────────────────────────────────────┐
│  STATUS │ ELAPSED │             PACKAGE              │ COVER │ PASS │ FAIL │ SKIP  │
│─────────┼─────────┼──────────────────────────────────┼───────┼──────┼──────┼───────│
│  FAIL   │  8.36s  │ github.com/cli/cli/v2/acceptance │  --   │  0   │  2   │  0    │
└────────────────────────────────────────────────────────────────────────────────────┘
```

And here we see the updated test passes against the branch changes:

```shell
‹wm/secret-env-var*›$ set -o pipefail && GH_ACCEPTANCE_SCRIPT=secret-require-remote-disambiguation.txtar GH_ACCEPTANCE_TOKEN=$(gh auth token --hostname github.com) GH_ACCEPTANCE_HOST=github.com GH_ACCEPTANCE_ORG=gh-acceptance-testing go test -tags acceptance -json -run ^TestSecrets$ github.com/cli/cli/v2/acceptance  | tparse --all go test
┌──────────────────────────────────────────────────────────────────────────────────────────────────────────┐
│  STATUS │ ELAPSED │                       TEST                       │             PACKAGE               │
│─────────┼─────────┼──────────────────────────────────────────────────┼───────────────────────────────────│
│  PASS   │    9.18 │ TestSecrets/secret-require-remote-disambiguation │ github.com/cli/cli/v2/acceptance  │
│  PASS   │    0.00 │ TestSecrets                                      │ github.com/cli/cli/v2/acceptance  │
└──────────────────────────────────────────────────────────────────────────────────────────────────────────┘
┌────────────────────────────────────────────────────────────────────────────────────┐
│  STATUS │ ELAPSED │             PACKAGE              │ COVER │ PASS │ FAIL │ SKIP  │
│─────────┼─────────┼──────────────────────────────────┼───────┼──────┼──────┼───────│
│  PASS   │  9.74s  │ github.com/cli/cli/v2/acceptance │  --   │  2   │  0   │  0    │
└────────────────────────────────────────────────────────────────────────────────────┘
```

andyfeller committed:
Fix gh secret disambiguity test

The previous commit failed to update the 2nd set of tests to rely upon GH_REPO instead of --repo flag.

These have been tested against `trunk` and fail because not being in a git directory when they should pass.

jtmcg commented on acceptance/testdata/secret/secret-require-remote-disambiguation.txtar:
```diff
@@ -34,3 +34,29 @@ stderr 'multiple remotes detected. please specify which repo to use by providing
 # Secret delete requires disambiguation
 ! exec gh secret delete 'TEST_SECRET_NAME'
 stderr 'multiple remotes detected. please specify which repo to use by providing the -R, --repo argument'
+
+# Move out of the fork repo to test whether secret commands work without local repository context
+cd ..
+
+# Secret set using --repo flag does not require disambiguation
+exec gh secret set 'TEST_SECRET_NAME' --body 'TEST_SECRET_VALUE' --repo ${ORG}/${REPO}-fork
+
+# Secret list using --repo flag does not require disambiguation
+exec gh secret list --repo ${ORG}/${REPO}-fork
+stdout 'TEST_SECRET_NAME'
+
+# Secret delete using --repo flag does not require disambiguation
+exec gh secret delete 'TEST_SECRET_NAME' --repo ${ORG}/${REPO}-fork
+
+# Setup GH_REPO for testing environment variable behavior
+env GH_REPO=${ORG}/${REPO}-fork
+
+# Secret set using GH_REPO env var does not require disambiguation
+exec gh secret set 'TEST_SECRET_NAME2' --body 'TEST_SECRET_VALUE2'
+
+# Secret list using --repo flag does not require disambiguation
+exec gh secret list
+stdout 'TEST_SECRET_NAME2'
+
+# Secret delete using --repo flag does not require disambiguation
+exec gh secret delete 'TEST_SECRET_NAME2'
```
Non-blocking: I count 6 added tests to this acceptance test (name on lines 41, 44, 48, 54, 57, & 61). What's the reasoning for not writing whole new acceptance test files for these?

andyfeller replied:
@jtmcg : Good question.

The existing acceptance tests exercise different secret types _(organization, repository, repository environment)_ with this acceptance test focused on `gh secret` resolving repository based on remotes.  Or atleast it does on the surface.

This acceptance test was created when the code under review here was modified.  In my mind, it is testing half of the problem _("Does `gh secret` warn users when there is ambiguity?")_ and these additions were adding the other half _("Does `gh secret` handle situations where it isn't ambiguous?")_.  So it felt like this would be a single place that handles exercising those concerns and helping to provide context if failures are reintroduced.

Otherwise, we can move these to a wholly new acceptance test.

jtmcg replied:
Yeah, I see what you mean. To take separating it out to the extreme, we could put all 6 of these into their own files. Alternatively, we could separate testing of `--repo` flag from `--env` flag. The reason I might advocate for either of these is that when any of these tests fail it gives would give us more granularity in debugging. As currently written, if the error case fails, we won't have insights into the other cases until that gets fixed because the test will exit there (I think).

That said, it's a tiny optimization, hence the "non-blocking" nature of this comment. I'll leave the decision to you.

jtmcg commented on pkg/cmd/secret/delete/delete_test.go:
```diff
@@ -144,6 +144,7 @@ func TestNewCmdDeleteBaseRepoFuncs(t *testing.T) {
 	tests := []struct {
 		name          string
 		args          string
+		env           map[string]string
```
What's the reasoning for using a `map[string]string` here and not just explicitly setting the `GH_REPO` env var like below?

```suggestion
		ghRepoEnv           string
```

andyfeller replied:
Truthfully, I was borrowing from similar patterns in other tests as well as the idea of putting the context in the testing table rather than the code.

Additionally, I think I was influenced by and earlier version of https://github.com/cli/cli/pull/10535 where the draft checked for the existing of `GH_REPO` environment variable rather than the value of it.  Thus this test is only setting environment variables and values intentionally.

andyfeller replied:
Other examples I was considering at the time:

https://github.com/cli/cli/blob/b8cf16dc2d4d8040470beba883847d002119925a/pkg/cmdutil/auth_check_test.go#L12-L57

https://github.com/cli/cli/blob/b8cf16dc2d4d8040470beba883847d002119925a/pkg/cmd/auth/status/status_test.go#L85-L560

https://github.com/cli/cli/blob/b8cf16dc2d4d8040470beba883847d002119925a/pkg/cmd/auth/switch/switch_test.go#L112-L380

jtmcg replied:
Yeah, I can see value in both approaches. The only issue is the generic `map[string]string` masks what env vars we'll actually be testing. This gets disambiguated for any test that requires env vars in the test definition, but not elsewhere.

Again, this is a tiny optimization, so I'll leave the decision to you 🙂

jtmcg commented on pkg/cmd/secret/delete/delete_test.go:
```diff
@@ -199,6 +208,10 @@ func TestNewCmdDeleteBaseRepoFuncs(t *testing.T) {
 				},
 			}
 
+			for k, v := range tt.env {
+				t.Setenv(k, v)
+			}
```
Following the above comment:

```suggestion
			t.Setenv("GH_REPO", tt.ghRepoEnvVar)
```

jtmcg reviewed (approved):
Looks good to me! I've got a couple questions, but they're non-blocking so I'm giving this my ✅

jtmcg commented on acceptance/testdata/secret/secret-require-remote-disambiguation.txtar:
```diff
@@ -34,3 +34,29 @@ stderr 'multiple remotes detected. please specify which repo to use by providing
+# Secret set using GH_REPO env var does not require disambiguation
+exec gh secret set 'TEST_SECRET_NAME2' --body 'TEST_SECRET_VALUE2'
+
+# Secret list using --repo flag does not require disambiguation
+exec gh secret list
+stdout 'TEST_SECRET_NAME2'
+
+# Secret delete using --repo flag does not require disambiguation
```
Just realized that these are referencing `--repo` and I think they should be referencing `GH_REPO` in the test descriptions

andyfeller committed:
Fix acceptance test descriptions
